File: java_solution/101. Symmetric Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 递归,Depth-first Search,Bloomberg,二叉树,领英,微软,Breadth-first Search
// Recursive
// Time complexity : O(n). Because we traverse the entire input tree once
// Space complexity : The number of recursive calls is bound by the height of the tree. In the worst case, the tree is linear and the height is in O(n).

class Solution {
    public boolean isSymmetric(TreeNode root) {
        return isMirror(root, root);
    }
    
    public boolean isMirror(TreeNode t1, TreeNode t2) {
        if (t1 == null && t2 == null) {
            return true;
        }
        
        if (t1 == null || t2 == null) {
            return false;
        }
        
        return (t1.val == t2.val) && isMirror(t1.right, t2.left) && isMirror(t1.left, t2.right);
        
    }
}

File: java_solution/101. Symmetric Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
 
class Solution:
    def _is_symmetric(self, l, r):
        """
        :type l: NodeTree
        :type r: NodeTree
        :rtype: bool
        """
        if not l and not r:
            return True
        if l and not r:
            return False
        if r and not l:
            return False
        if l.val != r.val:
            return False
        return self._is_symmetric(l.right, r.left) and self._is_symmetric(l.left, r.right)
        
    def isSymmetric(self, root):
        """
        :type root: NodeTree
        :rtype: bool
        """
        if not root:
            return True
        return self._is_symmetric(root.left, root.right)
 
